File: Problems/No.617_Merge_Two_Binary_Trees/merge_two_binary_trees.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def mergeTrees(self, root1, root2):
        """
        :type root1: TreeNode
        :type root2: TreeNode
        :rtype: TreeNode
        """

        # parameters
        TNodes = []
        curr_pair = []

        # start computing
        if (root1 == None):
            return root2
        elif (root2 == None):
            return root1
        else:
            TNodes.append([root1, root2])
            while TNodes:
                curr_pair = TNodes.pop(0)
                if ((curr_pair[0] == None) or (curr_pair[1] == None)):
                    continue
                else:
                    curr_pair[0].val += curr_pair[1].val
                    if curr_pair[0].left:
                        TNodes.append([curr_pair[0].left, curr_pair[1].left])
                    else:
                        curr_pair[0].left = curr_pair[1].left
                    if curr_pair[0].right:
                        TNodes.append([curr_pair[0].right, curr_pair[1].right])
                    else:
                        curr_pair[0].right = curr_pair[1].right
            return root1

File: Problems/No.617_Merge_Two_Binary_Trees/merge_two_binary_trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* mergeTrees(TreeNode* root1, TreeNode* root2) {
        
        // parameters
        std::queue<std::pair<TreeNode*, TreeNode*>> TNodes;
        std::pair<TreeNode*, TreeNode*> curr_pair;

        // start computing
        if (root1 == nullptr) return root2;
        else if (root2 == nullptr) return root1;
        else {

            TNodes.push(std::make_pair(root1, root2));
            while (!TNodes.empty()) {

                curr_pair = TNodes.front();
                TNodes.pop();

                if ((curr_pair.first == nullptr) || (curr_pair.second == nullptr)) continue;
                else {

                    curr_pair.first->val += curr_pair.second->val;

                    if (curr_pair.first->left) TNodes.push(std::make_pair(curr_pair.first->left, curr_pair.second->left));
                    else curr_pair.first->left = curr_pair.second->left;
                    if (curr_pair.first->right) TNodes.push(std::make_pair(curr_pair.first->right, curr_pair.second->right));
                    else curr_pair.first->right = curr_pair.second->right;

                }

            }

            return root1;

        }

    }
};
